src/coreclr/jit/lclmorph.cpp

```diff
@@ -1244,6 +1250,24 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
                 lclNode = indir->AsLclVarCommon();
                 break;
 
+            case IndirTransform::NarrowCast:
+                assert(varTypeIsIntegral(indir));
+                assert(varTypeIsIntegral(varDsc));
+                assert(*val.Use() == indir);
+                assert(!isDef);
+
+                lclNode    = BashToLclVar(indir->gtGetOp1(), lclNum);
+                *val.Use() = m_compiler->gtNewCastNode(genActualType(indir), lclNode, false, indir->TypeGet());
+
+                // Check to make sure this is narrow.
+                assert(genTypeSize(lclNode) >= genTypeSize(indir));
```

Comment:
It would be better to check `genTypeSize(varDsc) >= genTypeSize(indir)` (together with `varTypeIsIntegral(varDsc)`), since small `LCL_VAR`s can appear `TYP_INT`-typed.